helicone-node/async_logger/HeliconeManualLogger.ts

```diff
@@ -0,0 +1,159 @@
+export class HeliconeManualLogger {
+  private apiKey: string;
+  private headers: Record<string, string>;
+  private request: ILogRequest | null = null;
+
+  private startTime: number;
+  private endTime: number | null = null;
+
+  private readonly LOGGING_ENDPOINT: string = "https://api.hconeai.com/custom/v1/log";
+
+  constructor(opts: IHeliconeManualLogger) {
+    this.apiKey = opts.apiKey;
+    this.headers = opts.headers || {};
+    this.startTime = Date.now();
+  }
+
+  public registerRequest(request: ILogRequest): void {
+    if (request.messages) {
+
+    }
+    this.request = request;
+  }
+
+  public sendLog(response: ILogResponse, meta: Record<string, string>): void {
+    if (this.request === null) {
+      console.error("Request is not registered.");
+      return;
+    }
+
+    this.endTime = Date.now();
+
+    try {
+      const providerRequest: ProviderRequest = {
+        url: "custom-model-nopath",
+        json: this.request,
+        meta: meta ?? {}
+      }
+
+      const providerResponse: ProviderResponse = {
+        headers: this.headers,
+        status: parseInt(meta ? (meta.status ?? "200") : "200"),
+        json: response
+      }
+
+      const timing: Timing = {
+        startTime: {
+          seconds: Math.trunc(this.startTime / 1000),
+          milliseconds: this.startTime % 1000,
+        },
+        endTime: {
+          seconds: Math.trunc(this.endTime / 1000),
+          milliseconds: this.endTime % 1000,
+        },
+      }
+
+      const options = {
+        method: "POST",
+        headers: {
+          Authorization: `Bearer ${this.apiKey}`,
+          "Content-Type": "application/json",
+          ...this.headers,
+        },
+        body: JSON.stringify({
+          providerRequest,
+          providerResponse,
+          timing
+        }),
+      };
+
+      fetch(this.LOGGING_ENDPOINT, options);
+    } catch (error: any) {
+      console.error(
+        "Error making request to Helicone log endpoint:",
+        error?.message,
+        error
+      );
+
+      return;
+    }
+  }
+}
+
+
+/*
+ * Type Definitions
+ *
+ * */
+
+type ProviderRequest = {
+  url: string;
+  json: {
+    [key: string]: any;
+  };
+  meta: Record<string, string>;
+};
+
+type ProviderResponse = {
+  json: {
+    [key: string]: any;
+  };
+  status: number;
+  headers: Record<string, string>;
+};
+
+type Timing = {
+  startTime: {
+    seconds: number;
+    milliseconds: number;
+  };
+  endTime: {
+    seconds: number;
+    milliseconds: number;
+  };
+};
+
+type IHeliconeManualLogger = {
+  apiKey: string;
+  headers?: Record<string, string>;
+}
+
+type BaseRequest = {
+  model: string,
+  temperature?: number;
+  max_tokens?: number;
+  stream?: boolean;
+}
+
+type MessageBody = {
+  messages: Array<{
+    role: "system" | "user" | "assistant",
+    content: string
+  }>;
+}
+
+type StringBody = {
+  input: string;
+}
+
+type ILogRequest = BaseRequest & (MessageBody | StringBody);
```

Comment:
This type is useful while constructing the request body.

The request must contain _either_ `messages` or `input`, not both (nor none!).

```ts
let r: ILogRequest = {
    model: "gpt-3.5-turbo",
    temperature: 1,
    max_tokens: 50,
    stream: false,
    messages: [
        {
            role: "system",
            content: "you are a chatbot"
        }
    ],
    input: "hello"
};

console.log(r.messages);
console.log(r.input);
```

Here TS will complain when user tries to access one of the two since both were passed. If the user passed none of them, TS will again complain.

If only one of the properties is passed, then TS will give autocomplete for that property only.